// @swensonhe/strapi-plugin-firebase-auth/admin/src/pages/Settings/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var design_system_1 = require("@strapi/design-system");
var helper_plugin_1 = require("@strapi/helper-plugin");
var api_1 = require("./api");
var icons_1 = require("@strapi/icons");
var DeleteJsonConfigurationDialogue_1 = require("./DeleteJsonConfigurationDialogue");
var react_router_dom_1 = require("react-router-dom");
function SettingsPage() {
    var _this = this;
    var toggleNotification = (0, helper_plugin_1.useNotification)();
    var _a = __read((0, react_1.useState)(null), 2), firebaseJsonValue = _a[0], setFirebaseJsonValue = _a[1];
    var _b = __read((0, react_1.useState)(""), 2), firebaseJsonValueInput = _b[0], setFirebaseJsonValueInput = _b[1];
    var _c = __read((0, react_1.useState)(false), 2), isDeleteDialogueOpen = _c[0], setIsDeleteDialogueOpen = _c[1];
    var _d = __read((0, react_1.useState)(true), 2), loading = _d[0], setLoading = _d[1];
    var history = (0, react_router_dom_1.useHistory)();
    var handleRetrieveFirebaseJsonConfig = function () {
        setLoading(true);
        (0, api_1.getFirebaseConfig)()
            .then(function (data) {
            setLoading(false);
            setFirebaseJsonValue(data);
            setFirebaseJsonValueInput(data);
        })
            .catch(function (err) {
            setLoading(false);
        });
    };
    (0, react_1.useEffect)(function () {
        handleRetrieveFirebaseJsonConfig();
    }, []);
    var handleDeleteFirebaseJsonConfig = function () { return __awaiter(_this, void 0, void 0, function () {
        var err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    setLoading(true);
                    return [4, (0, api_1.delFirebaseConfig)()];
                case 1:
                    _a.sent();
                    setFirebaseJsonValue("");
                    setFirebaseJsonValueInput("");
                    setLoading(false);
                    toggleNotification({
                        type: "success",
                        message: "Firebase configuration has successfully been removed",
                    });
                    setIsDeleteDialogueOpen(false);
                    return [3, 3];
                case 2:
                    err_1 = _a.sent();
                    setLoading(false);
                    toggleNotification({
                        type: "warning",
                        message: "An error occured, please try again",
                    });
                    setIsDeleteDialogueOpen(false);
                    return [3, 3];
                case 3: return [2];
            }
        });
    }); };
    var handleFirebaseJsonSubmit = function () { return __awaiter(_this, void 0, void 0, function () {
        var data, error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    setLoading(true);
                    return [4, (0, api_1.saveFirebaseConfig)(firebaseJsonValueInput)];
                case 1:
                    data = _a.sent();
                    setFirebaseJsonValue(data["firebase-config-json"]);
                    setLoading(false);
                    toggleNotification({
                        type: "success",
                        message: {
                            id: "notification.success",
                            defaultMessage: "Data submitted successfully",
                        },
                    });
                    return [3, 3];
                case 2:
                    error_1 = _a.sent();
                    toggleNotification({
                        type: "warning",
                        message: {
                            id: "notification.error",
                            defaultMessage: "some thing went wrong",
                        },
                    });
                    setLoading(false);
                    return [3, 3];
                case 3: return [2];
            }
        });
    }); };
    if (loading) {
        return react_1.default.createElement(helper_plugin_1.LoadingIndicatorPage, null);
    }
    var isJsonString = function (str) {
        try {
            JSON.parse(str);
        }
        catch (e) {
            return false;
        }
        return true;
    };
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(DeleteJsonConfigurationDialogue_1.DeleteJsonConfigurationDialogue, { isOpen: isDeleteDialogueOpen, onToggleDialog: function () {
                setIsDeleteDialogueOpen(false);
            }, onConfirm: handleDeleteFirebaseJsonConfig }),
        react_1.default.createElement(design_system_1.Flex, { style: { padding: 32 }, direction: "column", alignItems: "flex-start", gap: 4 },
            react_1.default.createElement(design_system_1.Box, { style: { width: "100%" } },
                !firebaseJsonValue ? (react_1.default.createElement(react_1.default.Fragment, null,
                    react_1.default.createElement(design_system_1.JSONInput, { label: "Firebase-json-configuration", value: firebaseJsonValueInput, height: 500, style: { height: 400 }, onChange: setFirebaseJsonValueInput, error: firebaseJsonValueInput &&
                            !isJsonString(firebaseJsonValueInput)
                            ? "Please enter a valid JSON string"
                            : "" }),
                    react_1.default.createElement(design_system_1.Flex, { style: {
                            marginTop: 32,
                            width: "100%",
                            padding: 16,
                        }, justifyContent: "flex-end" },
                        react_1.default.createElement(design_system_1.Button, { size: "L", onClick: handleFirebaseJsonSubmit, disabled: !isJsonString(firebaseJsonValueInput) }, "Submit")))) : (react_1.default.createElement(react_1.default.Fragment, null,
                    react_1.default.createElement(design_system_1.Flex, { gap: 4 },
                        "You have successfully submitted your json configuration for project:",
                        " ",
                        react_1.default.createElement("span", { style: { fontWeight: 700 } }, (firebaseJsonValue &&
                            JSON.parse(firebaseJsonValue.firebaseConfigJson)
                                .project_id) ||
                            JSON.parse(firebaseJsonValue.firebaseConfigJson).projectId),
                        react_1.default.createElement("button", { onClick: function () {
                                setIsDeleteDialogueOpen(true);
                            } },
                            react_1.default.createElement(icons_1.Trash, null))),
                    react_1.default.createElement(design_system_1.Button, { onClick: function () { return history.push("/plugins/firebase-auth"); }, marginTop: 4 }, "Back to firebase plugin"))),
                !firebaseJsonValue ? (react_1.default.createElement(design_system_1.Flex, { direction: "column", alignItems: "flex-start", marginTop: 10 },
                    "How to setup the firebase configuration:",
                    react_1.default.createElement(design_system_1.Box, { marginTop: 2, marginLeft: 6 },
                        react_1.default.createElement("ol", { style: { listStyle: "auto" } },
                            react_1.default.createElement("li", { style: { marginTop: 32 } },
                                react_1.default.createElement(design_system_1.Typography, null,
                                    "Go to your Project",
                                    " ",
                                    react_1.default.createElement("a", { href: "https://console.firebase.google.com/project/_/settings/general/", target: "_blank", rel: "noreferrer" }, "settings"),
                                    " ",
                                    "in the Firebase console.")),
                            react_1.default.createElement("li", { style: { marginTop: 32 } },
                                react_1.default.createElement(design_system_1.Typography, null, "Select the project in your list of projects")),
                            react_1.default.createElement("li", { style: { marginTop: 32 } },
                                react_1.default.createElement(design_system_1.Typography, null, "Scroll down to the apps section and select your app"),
                                react_1.default.createElement("img", { src: require("../../assets/firebase-config-tutorial/select-app.png"), style: { maxWidth: "100%", marginTop: 32 } })),
                            react_1.default.createElement("li", { style: { marginTop: 32 } },
                                react_1.default.createElement(design_system_1.Typography, null, "Select Config from the Firebase SDK snippet pane and copy the JSON object."),
                                react_1.default.createElement("div", null,
                                    react_1.default.createElement("img", { src: require("../../assets/firebase-config-tutorial/config.png"), style: { maxWidth: "100%", marginTop: 32 } }))),
                            react_1.default.createElement("li", { style: { marginTop: 32 } },
                                react_1.default.createElement(design_system_1.Typography, null, "Stringify the JSON object and Paste it in the Firebase-json-configuration input field above and click submit."),
                                react_1.default.createElement("div", { style: { marginTop: 24 } },
                                    react_1.default.createElement("img", { src: require("../../assets/firebase-config-tutorial/submission.png"), style: { maxWidth: "100%" } }))))))) : null))));
}
exports.default = SettingsPage;
