// @swensonhe/strapi-plugin-firebase-auth/admin/src/pages/HomePage/HomePage.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HomePage = void 0;
var react_1 = __importStar(require("react"));
var react_helmet_1 = require("react-helmet");
var helper_plugin_1 = require("@strapi/helper-plugin");
var design_system_1 = require("@strapi/design-system");
var design_system_2 = require("@strapi/design-system");
var design_system_3 = require("@strapi/design-system");
var design_system_4 = require("@strapi/design-system");
var react_query_1 = require("react-query");
var api_1 = require("./utils/api");
var ListView_1 = __importDefault(require("../ListView"));
var icons_1 = require("@strapi/icons");
var react_router_dom_1 = require("react-router-dom");
var api_2 = require("../Settings/api");
var INITIAL_USERS_DATA = {
    data: [],
    meta: { pagination: { page: 0, pageCount: 0, pageSize: 0, total: 0 } },
};
var HomePage = function () {
    var toggleNotification = (0, helper_plugin_1.useNotification)();
    var _a = __read((0, react_1.useState)(INITIAL_USERS_DATA), 2), usersData = _a[0], setUsersData = _a[1];
    var _b = __read((0, react_1.useState)(false), 2), isNotConfigured = _b[0], setIsNotConfigured = _b[1];
    var _c = __read((0, react_1.useState)(true), 2), isLoadingConfiguration = _c[0], setIsLoadingConfiguration = _c[1];
    var history = (0, react_router_dom_1.useHistory)();
    var handleRetrieveFirebaseJsonConfig = function () {
        (0, api_2.getFirebaseConfig)()
            .then(function () {
            setIsNotConfigured(false);
            setIsLoadingConfiguration(false);
        })
            .catch(function (err) {
            setIsNotConfigured(true);
            setIsLoadingConfiguration(false);
        });
    };
    (0, react_1.useEffect)(function () {
        handleRetrieveFirebaseJsonConfig();
    }, []);
    var isLoading = (0, react_query_1.useQuery)("firebase-auth-", function () { return (0, api_1.fetchUsers)(); }, {
        onSuccess: function (result) {
            setUsersData(result);
        },
        onError: function (err) {
            toggleNotification({
                type: "warning",
                message: {
                    id: "notification.error",
                    defaultMessage: "An error occured",
                },
            });
        },
        enabled: !isLoadingConfiguration && !isNotConfigured,
    }).isLoading;
    if (isLoading || isLoadingConfiguration) {
        return react_1.default.createElement(helper_plugin_1.LoadingIndicatorPage, null);
    }
    return (react_1.default.createElement(design_system_1.Layout, null,
        react_1.default.createElement(react_helmet_1.Helmet, { title: "Firebase Users" }),
        react_1.default.createElement(design_system_2.Main, null,
            react_1.default.createElement(design_system_3.Box, { padding: 10 }, !isNotConfigured ? (react_1.default.createElement(design_system_4.Grid, { gap: 4 },
                react_1.default.createElement(design_system_4.GridItem, { col: 12, s: 12 },
                    react_1.default.createElement(ListView_1.default, { data: usersData.data, meta: usersData.meta })))) : (react_1.default.createElement(design_system_4.Flex, { direction: "column", marginTop: 10 },
                react_1.default.createElement(icons_1.ExclamationMarkCircle, null),
                react_1.default.createElement(design_system_3.Box, { marginTop: 1 },
                    react_1.default.createElement(design_system_4.Typography, null, "Firebase is not configured, please configure Firebase")),
                react_1.default.createElement(design_system_4.Button, { marginTop: 3, onClick: function () {
                        history.push("/settings/firebase-auth");
                    } }, "Configure firebase")))))));
};
exports.HomePage = HomePage;
