// @swensonhe/strapi-plugin-firebase-auth/admin/src/pages/EditView/EditView.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditView = void 0;
var react_1 = __importDefault(require("react"));
var react_helmet_1 = require("react-helmet");
var design_system_1 = require("@strapi/design-system");
var react_router_dom_1 = require("react-router-dom");
var api_1 = require("../HomePage/utils/api");
var react_query_1 = require("react-query");
var helper_plugin_1 = require("@strapi/helper-plugin");
var EditForm_1 = require("./EditForm");
var EditView = function () {
    var id = (0, react_router_dom_1.useParams)().id;
    var toggleNotification = (0, helper_plugin_1.useNotification)();
    var _a = (0, react_query_1.useQuery)("firebase-auth-".concat(id), function () { return (0, api_1.fetchUserByID)(id); }, {
        onError: function () {
            toggleNotification({
                type: "warning",
                message: {
                    id: "notification.error",
                    defaultMessage: "An error occured",
                },
            });
        },
    }), status = _a.status, data = _a.data;
    var isLoadingUsersData = status !== "success" && status !== "error";
    if (isLoadingUsersData) {
        return react_1.default.createElement(helper_plugin_1.LoadingIndicatorPage, null);
    }
    return (react_1.default.createElement(design_system_1.Layout, null,
        react_1.default.createElement(react_helmet_1.Helmet, { title: "Firebase User" }),
        react_1.default.createElement(EditForm_1.EditForm, { data: data })));
};
exports.EditView = EditView;
