// @swensonhe/strapi-plugin-firebase-auth/admin/src/pages/EditView/EditForm.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditForm = void 0;
var react_1 = __importStar(require("react"));
var design_system_1 = require("@strapi/design-system");
var design_system_2 = require("@strapi/design-system");
var design_system_3 = require("@strapi/design-system");
var design_system_4 = require("@strapi/design-system");
var design_system_5 = require("@strapi/design-system");
var design_system_6 = require("@strapi/design-system");
var design_system_7 = require("@strapi/design-system");
var Header_1 = require("../Header/Header");
var design_system_8 = require("@strapi/design-system");
var helper_plugin_1 = require("@strapi/helper-plugin");
var design_system_9 = require("@strapi/design-system");
var icons_1 = require("@strapi/icons");
var design_system_10 = require("@strapi/design-system");
var styled_components_1 = __importDefault(require("styled-components"));
var date_fns_1 = require("date-fns");
var api_1 = require("../HomePage/utils/api");
var design_system_11 = require("@strapi/design-system");
var react_router_dom_1 = require("react-router-dom");
var USERS_URL = "/content-manager/collectionType/plugin::users-permissions.user";
var ContentWrapper = (0, styled_components_1.default)(design_system_5.Box)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  width: 100%;\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  font-size: 18px;\n  padding: 5px;\n"], ["\n  width: 100%;\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  font-size: 18px;\n  padding: 5px;\n"])));
var DetailsButtonWrapper = (0, styled_components_1.default)(design_system_5.Box)(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  width: 100%;\n  display: flex;\n  justify-content: flex-end;\n  align-items: center;\n  padding: 5px;\n"], ["\n  width: 100%;\n  display: flex;\n  justify-content: flex-end;\n  align-items: center;\n  padding: 5px;\n"])));
var MetaWrapper = (0, styled_components_1.default)(design_system_5.Box)(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  width: 100%;\n  display: flex;\n  flex-direction: column;\n  justify-content: flex-start;\n  font-size: 18px;\n  padding: 5px;\n"], ["\n  width: 100%;\n  display: flex;\n  flex-direction: column;\n  justify-content: flex-start;\n  font-size: 18px;\n  padding: 5px;\n"])));
var EditForm = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j;
    var data = _a.data;
    var _k = __read((0, react_1.useState)(data), 2), userData = _k[0], setUserData = _k[1];
    var _l = __read((0, react_1.useState)(data), 2), originalUserData = _l[0], setOriginalUserData = _l[1];
    var _m = __read((0, react_1.useState)(false), 2), isLoading = _m[0], setIsLoading = _m[1];
    var toggleNotification = (0, helper_plugin_1.useNotification)();
    var location = (0, react_router_dom_1.useLocation)();
    var history = (0, react_router_dom_1.useHistory)();
    var locationState = location.state;
    var onTextInputChange = function (e) {
        e.preventDefault();
        setUserData(function (prevState) {
            var _a;
            return (__assign(__assign({}, prevState), (_a = {}, _a[e.target.name] = e.target.value, _a)));
        });
    };
    var onToggleInputChange = function (e) {
        setUserData(function (prevState) {
            var _a;
            return (__assign(__assign({}, prevState), (_a = {}, _a[e.target.name] = e.target.checked, _a)));
        });
    };
    var updateUserHandler = function () { return __awaiter(void 0, void 0, void 0, function () {
        var updatedUser_1, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    setIsLoading(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4, (0, api_1.updateUser)(userData.uid, userData)];
                case 2:
                    updatedUser_1 = _a.sent();
                    if (updatedUser_1[0].status === "rejected")
                        throw new Error("Error updating user");
                    setUserData(function () { return updatedUser_1[0].value; });
                    setOriginalUserData(function () { return updatedUser_1[0].value; });
                    setIsLoading(false);
                    toggleNotification({
                        type: "success",
                        message: { id: "notification.success", defaultMessage: "Saved" },
                    });
                    return [3, 4];
                case 3:
                    e_1 = _a.sent();
                    console.log("err --> ", e_1);
                    toggleNotification({
                        type: "warning",
                        message: {
                            id: "notification.error",
                            defaultMessage: "An error occured!",
                        },
                    });
                    setIsLoading(false);
                    setUserData(data);
                    return [3, 4];
                case 4: return [2];
            }
        });
    }); };
    if (isLoading) {
        return react_1.default.createElement(helper_plugin_1.LoadingIndicatorPage, null);
    }
    return (react_1.default.createElement(design_system_2.Main, null,
        react_1.default.createElement(Header_1.Header, { title: "Edit User", onSave: updateUserHandler, initialData: originalUserData, modifiedData: userData, isSubmitButtonDisabled: !(userData === null || userData === void 0 ? void 0 : userData.email) ||
                !!(((_b = userData === null || userData === void 0 ? void 0 : userData.password) === null || _b === void 0 ? void 0 : _b.length) && ((_c = userData === null || userData === void 0 ? void 0 : userData.password) === null || _c === void 0 ? void 0 : _c.length) < 6) }),
        react_1.default.createElement(design_system_4.ContentLayout, null,
            react_1.default.createElement(design_system_3.Grid, { gap: 4 },
                react_1.default.createElement(design_system_3.GridItem, { col: 9, s: 12 },
                    react_1.default.createElement(design_system_5.Box, { background: "neutral0", borderColor: "neutral150", hasRadius: true },
                        react_1.default.createElement(design_system_1.Stack, { spacing: 2, padding: 3, gap: 2 },
                            react_1.default.createElement(design_system_6.TextInput, { id: "email", name: "email", autoComplete: "new-password", onChange: onTextInputChange, label: "Email", value: userData.email, error: !(userData === null || userData === void 0 ? void 0 : userData.email) ? "Email is required" : "", required: true }),
                            react_1.default.createElement(design_system_6.TextInput, { id: "displayName", name: "displayName", autoComplete: "new-password", onChange: onTextInputChange, label: "Display Name", value: userData.displayName }),
                            react_1.default.createElement(design_system_6.TextInput, { id: "phoneNumber", name: "phoneNumber", autoComplete: "new-password", onChange: onTextInputChange, label: "Phone Number", value: userData.phoneNumber }),
                            react_1.default.createElement(design_system_6.TextInput, { id: "password", name: "password", type: "password", autoComplete: "new-password", onChange: onTextInputChange, label: "Password", value: userData.password, error: ((_d = userData === null || userData === void 0 ? void 0 : userData.password) === null || _d === void 0 ? void 0 : _d.length) && ((_e = userData === null || userData === void 0 ? void 0 : userData.password) === null || _e === void 0 ? void 0 : _e.length) < 6
                                    ? "Password must be at least 6 characters"
                                    : "" }),
                            react_1.default.createElement(design_system_8.ToggleInput, { label: "Disabled", name: "disabled", onLabel: "True", offLabel: "False", checked: Boolean(userData.disabled), onChange: onToggleInputChange }),
                            react_1.default.createElement(design_system_8.ToggleInput, { label: "Email Verified", name: "emailVerified", onLabel: "True", offLabel: "False", checked: Boolean(userData.emailVerified), onChange: onToggleInputChange })))),
                react_1.default.createElement(design_system_3.GridItem, { col: 3, s: 12 },
                    react_1.default.createElement(design_system_5.Box, { as: "aside", "aria-labelledby": "additional-informations", background: "neutral0", borderColor: "neutral150", hasRadius: true, paddingBottom: 2, paddingLeft: 4, paddingRight: 4, paddingTop: 2, shadow: "tableShadow" }, (_f = userData.providerData) === null || _f === void 0 ? void 0 :
                        _f.map(function (provider) {
                            return (react_1.default.createElement(design_system_11.Flex, { paddingTop: 2, paddingBottom: 2, direction: "column", alignItems: "flex-start", gap: 2 },
                                react_1.default.createElement(design_system_11.Flex, { gap: 1 },
                                    react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, "Provider Id:"),
                                    react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, provider.providerId)),
                                react_1.default.createElement(design_system_11.Flex, { gap: 1 },
                                    react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, "UID:"),
                                    react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, provider.uid)),
                                locationState.strapiId ? (react_1.default.createElement(design_system_11.Flex, { gap: 1 },
                                    react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, "Strapi ID:"),
                                    react_1.default.createElement(design_system_10.Typography, { variant: "sigma", id: "relations-title" },
                                        react_1.default.createElement("button", { style: {
                                                color: "blue",
                                                textDecoration: "underline",
                                            }, onClick: function () {
                                                history.push("/content-manager/collectionType/plugin::users-permissions.user/".concat(locationState.strapiId));
                                                return false;
                                            } }, locationState.strapiId)))) : null));
                        }),
                        react_1.default.createElement(design_system_9.Divider, null),
                        react_1.default.createElement(design_system_11.Flex, { paddingTop: 2, paddingBottom: 2, direction: "column", alignItems: "flex-start" },
                            ((_g = userData.metadata) === null || _g === void 0 ? void 0 : _g.lastSignInTime) && (react_1.default.createElement(MetaWrapper, null,
                                react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, "Last Sign In Time"),
                                react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, (0, date_fns_1.format)(new Date(userData.metadata.lastSignInTime), "yyyy/MM/dd kk:mm")))),
                            ((_h = userData.metadata) === null || _h === void 0 ? void 0 : _h.creationTime) && (react_1.default.createElement(MetaWrapper, null,
                                react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, "Creation Time"),
                                react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, (0, date_fns_1.format)(new Date((_j = userData.metadata) === null || _j === void 0 ? void 0 : _j.creationTime), "yyyy/MM/dd kk:mm")))))),
                    react_1.default.createElement(design_system_5.Box, { marginTop: 5, marginBottom: 5 }),
                    userData.localUser && (react_1.default.createElement(design_system_5.Box, { as: "aside", "aria-labelledby": "additional-informations", background: "neutral0", borderColor: "neutral150", hasRadius: true, paddingBottom: 1, paddingLeft: 2, paddingRight: 2, paddingTop: 1, shadow: "tableShadow" },
                        react_1.default.createElement(design_system_5.Box, { paddingTop: 2, paddingBottom: 2 },
                            react_1.default.createElement(DetailsButtonWrapper, null,
                                react_1.default.createElement(design_system_7.Link, { size: "S", startIcon: react_1.default.createElement(icons_1.Pencil, null), to: "".concat(USERS_URL, "/").concat(userData.localUser.id) }, "Details")),
                            react_1.default.createElement(ContentWrapper, { label: "UID" },
                                react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, "local user:"),
                                react_1.default.createElement(design_system_10.Typography, { variant: "sigma", textColor: "neutral600", id: "relations-title" }, userData.localUser.username))))))))));
};
exports.EditForm = EditForm;
var templateObject_1, templateObject_2, templateObject_3;
